src/hotspot/share/nmt/memoryFileTracker.cpp

```diff
@@ -0,0 +1,177 @@
+/*
+ * Copyright (c) 2024, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+#include "precompiled.hpp"
+#include "memory/allocation.hpp"
+#include "nmt/nmtCommon.hpp"
+#include "nmt/memoryFileTracker.hpp"
+#include "nmt/nmtNativeCallStackStorage.hpp"
+#include "nmt/vmatree.hpp"
+#include "runtime/mutex.hpp"
+#include "utilities/growableArray.hpp"
+#include "utilities/nativeCallStack.hpp"
+#include "utilities/ostream.hpp"
+
+MemoryFileTracker* MemoryFileTracker::Instance::_tracker = nullptr;
+PlatformMutex* MemoryFileTracker::Instance::_mutex = nullptr;
+
+MemoryFileTracker::MemoryFileTracker(bool is_detailed_mode)
+  : _stack_storage(is_detailed_mode), _devices() {}
+
+void MemoryFileTracker::allocate_memory(MemoryFile* device, size_t offset,
+                                        size_t size, const NativeCallStack& stack,
+                                        MEMFLAGS flag) {
+  NativeCallStackStorage::StackIndex sidx = _stack_storage.push(stack);
+  VMATree::RegionData regiondata(sidx, flag);
+  VMATree::SummaryDiff diff = device->_tree.reserve_mapping(offset, size, regiondata);
+  for (int i = 0; i < mt_number_of_types; i++) {
+    VirtualMemory* summary = device->_summary.by_type(NMTUtil::index_to_flag(i));
+    summary->reserve_memory(diff.flag[i].reserve);
```

Comment:
Why do we only track reserved memory here?

Reply:
This is clearly a point of confusion, as Gerard also asked about this. The answer is that the MFT only cares about memory in a file, which is always considered committed... So we consider reserved memory to be committed. Yeah, let's just change it so that MFT always commits memory instead.

Reply:
For the purpose of consistency with the rest of NMT, and to be able to use normal VirtualSummary etc, maybe MFT should mark the whole range as reserved upfront?

Reply:
I don't think that's the right choice here as the typical use-case is probably that you:

1. Reserve memory in the virtual address space
2. Make a memory file
3. Map in memory from that file to the memory you previously reserved

So if the memory file is also reserved, then that would cause "double accounting" of sorts. I'm also not sure if you have equivalent semantics of reserved memory for temporary files (doesn't actually take up space, is non-writeable).